R/parse.r

```diff
@@ -12,107 +12,85 @@ parse_all <- function(x) UseMethod("parse_all")
 
 #' @export
 parse_all.character <- function(x) {
-  x <- unlist(str_split(x, "\n"), recursive = FALSE, use.names = FALSE)
+
+  op <- options(stringsAsFactors = FALSE)
+  on.exit(options(op), add = TRUE)
+
+  if (length(grep("\n", x)))
+    x <- unlist(str_split(x, "\n"), recursive = FALSE, use.names = FALSE)
+  n <- length(x)
   src <- srcfilecopy("<text>", x)
+  exprs <- parse(text = x, srcfile = src)
 
-  expr <- parse(text = x, srcfile = src)
-  # No code, all comments
-  if (length(expr) == 0) {
-    n <- length(x)
-    if (n > 1) x <- paste(x, rep(c("\n", ""), c(n - 1, 1)), sep = "")
-    return(data.frame(
-      x1 = seq_along(x), x2 = seq_along(x),
-      y1 = rep(0, n), y2 = nchar(x),
-      src = x, text = rep(TRUE, n),
-      expr = I(rep(list(NULL), n)), visible = rep(FALSE, n),
-      stringsAsFactors = FALSE
-    ))
+  # No code, only comments and/or empty lines
+  ne <- length(exprs)
+  if (ne == 0) {
+    return(data.frame(src = append_break(x), expr = I(rep(list(NULL), n))))
   }
 
-  srcref <- attr(expr, "srcref")
-  srcfile <- attr(srcref[[1]], "srcfile")
-
-  # Create data frame containing each expression and its
-  # location in the original source
-  src <- sapply(srcref, function(src) str_c(as.character(src), collapse="\n"))
-  pos <- t(sapply(srcref, unclass))[, 1:4, drop = FALSE]
-  colnames(pos) <- c("x1", "y1", "x2", "y2")
-  pos <- as.data.frame(pos)[c("x1","y1","x2","y2")]
-
-  parsed <- data.frame(
-    pos, src=src, expr=I(as.list(expr)), text = FALSE,
-    stringsAsFactors = FALSE
-  )
-  # Extract unparsed text ----------------------------------------------------
-  # Unparsed text includes:
-  #  * text before first expression
-  #  * text between expressions
-  #  * text after last expression
-  #
-  # Unparsed text does not contain any expressions, so can
-  # be split into individual lines
-
-  get_region <- function(x1, y1, x2, y2) {
-    string <- getSrcRegion(srcfile, x1, x2, y1, y2)
-    lines <- strsplit(string, "(?<=\n)", perl=TRUE)[[1]]
-    n <- length(lines)
-    if (n == 0) {
-      lines <- ""
-      n <- 1
-    }
+  srcref <- attr(exprs, "srcref", exact = TRUE)
 
+  # Stard/End line numbers of expressions
+  pos <- as.data.frame(t(sapply(srcref, unclass))[, c(1, 3), drop = FALSE])
```

Comment:
`sapply()` is dangerous - maybe it could be replaced just with `lapply()` ? (or is that in the wrong direction?)